### PR TITLE
Threads: added support to read receipts (MSC3771)

### PR DESCRIPTION
**Pull request by gileluard**
needed for https://github.com/vector-im/element-ios/issues/6663

### CODE REVIEW

File: MatrixSDK/JSONModels/MXEvent.m [hunk 1]
```diff
@@ -965,6 +1005,11 @@ - (NSString *)threadId
     return nil;
 }
 
+//- (NSString *)threadIdForReadReceipt
```
Comment (ismailgulek) [inline MatrixSDK/JSONModels/MXEvent.m:1008]: Can we remove this property if not used?

File: MatrixSDK/Data/Store/MXFileStore/MXFileStore.m [hunk 2]
```diff
@@ -956,55 +957,101 @@ - (MXMemoryRoomOutgoingMessagesStore *)getOrCreateRoomOutgoingMessagesStore:(NSS
     return store;
 }
 
-- (RoomReceiptsStore*)getOrCreateRoomReceiptsStore:(NSString *)roomId
-{
-    RoomReceiptsStore *receiptsStore = roomReceiptsStores[roomId];
-    if (nil == receiptsStore)
+//- (RoomReceiptsStore*)getOrCreateRoomReceiptsStore:(NSString *)roomId
```
Comment (ismailgulek) [inline MatrixSDK/Data/Store/MXFileStore/MXFileStore.m:960]: Can we clean this?

File: MatrixSDK/Data/Store/MXMemoryStore/MXMemoryStore.h [hunk 3]
```diff
@@ -78,6 +83,23 @@ typedef NSMutableDictionary<NSString*, MXReceiptData*> RoomReceiptsStore;
  @param roomId the id of the room.
  @return receipts dictionary by user id.
  */
-- (RoomReceiptsStore*)getOrCreateRoomReceiptsStore:(NSString*)roomId;
+//- (RoomReceiptsStore*)getOrCreateRoomReceiptsStore:(NSString*)roomId;
```
Comment (ismailgulek) [inline MatrixSDK/Data/Store/MXMemoryStore/MXMemoryStore.h:86]: Can we remove this old method?

File: MatrixSDK/JSONModels/MXEvent.h [hunk 4]
```diff
@@ -645,7 +656,13 @@ extern NSString *const kMXEventIdentifierKey;
...
  nil if the event is not in a thread.
  */
-@property (nonatomic, readonly) NSString *threadId;
+@property (nonatomic, readonly, nullable) NSString *threadId;
+
+/**
+ Thread id for the event. This is actually the eventId of the thread's root event.
+ `main` if the event is not in a thread.
+ */
+//@property (nonatomic, readonly) NSString *threadIdForReadReceipt;
```
Comment (ismailgulek) [inline MatrixSDK/JSONModels/MXEvent.h:665]: Can we remove this?

File: MatrixSDKTests/MXReceiptDataIntegrationTests.swift [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+    
+    /// Test: Expect a thread is listed in the thread list after sending an event to a thread
+    /// - Create a Bob session
+    /// - Create an initial room
+    /// - Send a text message A to be used as thread root event
+    /// - Send a threaded event B referencing the root event A
+    /// - Expect a thread created with identifier A
+    /// - Expect thread's last message is B
+    /// - Expect thread has the root event
+    /// - Expect thread's number of replies is 1
```
Comment (ismailgulek) [inline MatrixSDKTests/MXReceiptDataIntegrationTests.swift:47]: These seem old. You can remove them if you don't want to describe them.

File: MatrixSDK/Contrib/Swift/MXRestClient.swift [hunk 6]
```diff
@@ -1735,8 +1735,8 @@ public extension MXRestClient {
      
      - returns: a `MXHTTPOperation` instance.
      */
-    @nonobjc @discardableResult func sendReadReceipt(toRoom roomId: String, forEvent eventId: String, completion: @escaping (_ response: MXResponse<Void>) -> Void) -> MXHTTPOperation {
-        return __sendReadReceipt(roomId, eventId: eventId, success: currySuccess(completion), failure: curryFailure(completion))
+    @nonobjc @discardableResult func sendReadReceipt(toRoom roomId: String, forEvent eventId: String, threadId: String?, completion: @escaping (_ response: MXResponse<Void>) -> Void) -> MXHTTPOperation {
```
Comment (ismailgulek) [inline MatrixSDK/Contrib/Swift/MXRestClient.swift:1738]: Can we add a doc line for the new parameter?

File: MatrixSDK/Data/MXRoom.h [hunk 7]
```diff
@@ -1309,7 +1311,7 @@ Remove a tag applied on an event of the room
  @param ts the receipt timestamp in ms since Epoch.
  @return YES if the receipt data is valid and has been stored.
  */
-- (BOOL)storeLocalReceipt:(NSString*)receiptType eventId:(NSString*)eventId userId:(NSString*)userId ts:(uint64_t)ts;
+- (BOOL)storeLocalReceipt:(NSString*)receiptType eventId:(NSString*)eventId threadId:(nullable NSString*)threadId userId:(NSString*)userId ts:(uint64_t)ts;
```
Comment (ismailgulek) [inline MatrixSDK/Data/MXRoom.h:1314]: Can we add a doc line for the new parameter?

**Review by ismailgulek — COMMENTED** [review]
Had some inline comments, otherwise lgtm

File: MatrixSDK/JSONModels/MXEvent.m [hunk 8]
```diff
@@ -574,6 +578,42 @@ - (NSArray *)readReceiptEventIds
...
+                for (NSDictionary *userDict in userDicts)
+                {
+                    threadId = userDict[@"thread_id"];
+                    if (threadId)
+                    {
+                        break;
+                    }
+                }
+                
+                [list addObject:threadId ?: [NSNull null]];
```
Comment (ismailgulek) [inline MatrixSDK/JSONModels/MXEvent.m:609]: Why don't we use `kMXEventTimelineMain` instead of `NSNull`?

Comment (gileluard) [reply MatrixSDK/JSONModels/MXEvent.m:609]: We still must support RR without thread ID (unthreaded RR) sent by old clients. In this case we have to set the thread ID as `nil` so it the RR will be stored in all thread and main timelines as per MSC.

File: MatrixSDK/Data/MXRoom.h [hunk 9]
```diff
@@ -1291,10 +1291,12 @@ Remove a tag applied on an event of the room
  Returns the read receipts list for an event, excluding the read receipt from the current user.
 
  @param eventId The event Id.
+ @param threadId The thread Id. Use `kMXEventTimelineMain` for the main timeline.
  @param sort YES to sort them from the latest to the oldest.
  @param completion Completion block containing the receipts for an event in a dedicated room.
  */
 - (void)getEventReceipts:(nonnull NSString*)eventId
+                threadId:(nullable NSString*)threadId
```
Comment (ismailgulek) [inline MatrixSDK/Data/MXRoom.h:1299]: Does it make sense to make this threadId `nonnull` now, as we expect `kMXEventTimelineMain` for the main timeline?

**Review by ismailgulek — APPROVED** [review]
A tiny comment inline, otherwise LGTM!

**Review by ismailgulek — APPROVED** [review]

**Pull request merged**